src/main/java/net/minecraftforge/client/LevelRendererHooks.java

```diff
@@ -0,0 +1,131 @@
+        if (!hooks.isEmpty())
+        {
+            Minecraft.getInstance().getProfiler().popPush(phase.name);
+            RenderContext ctx = new RenderContext(levelRenderer, poseStack, projectionMatrix, ticks, partialTicks, camX, camY, camZ);
+            for (Consumer<RenderContext> hook : hooks)
+                hook.accept(ctx);
+        }
+    }
+
+    /**
+     * Renders the registered for the {@link Phase} that matches the RenderType passed, if a Phase exists.
+     */
+    public static void render(RenderType renderType, LevelRenderer levelRenderer, PoseStack poseStack, Matrix4f projectionMatrix, int ticks, double camX, double camY, double camZ)
+    {
+        Phase phase = RENDER_TYPE_PHASES.get(renderType);
+        if (phase != null)
+            render(phase, levelRenderer, poseStack, projectionMatrix, ticks, camX, camY, camZ);
+    }
+
+    /**
+     * Times during {@link LevelRenderer#renderLevel} to render hooks.
+     */
+    public static class Phase
+    {
+        /** Called regardless of if they sky actually renders or not. */
+        public static final Phase AFTER_SKY = create("after_sky", null);
+        public static final Phase AFTER_SOLID_BLOCKS = create("after_solid_blocks", RenderType.solid());
+        public static final Phase AFTER_CUTOUT_MIPPED_BLOCKS_BLOCKS = create("after_cutout_mipped_blocks", RenderType.cutoutMipped());
+        public static final Phase AFTER_CUTOUT_BLOCKS = create("after_cutout_blocks", RenderType.cutout());
+        public static final Phase AFTER_ENTITIES = create("after_entities", null);
+        /** Called within a fabulous graphics target. Also happens after block entities */
+        public static final Phase AFTER_TRANSLUCENT_BLOCKS = create("after_translucent_blocks", RenderType.translucent());
+        public static final Phase AFTER_TRIPWIRE_BLOCKS = create("after_tripwire_blocks", RenderType.tripwire());
+        /** Called within a fabulous graphics target. */
+        public static final Phase AFTER_PARTICLES = create("after_particles", null);
+        /** Called within a fabulous graphics target. Only renders when clouds actually render. */
+        public static final Phase AFTER_CLOUDS = create("after_clouds", null);
+        /** Called within a fabulous graphics target. */
+        public static final Phase AFTER_WEATHER = create("after_weather", null);
+        /** May not work with fabulous graphics. */
+        public static final Phase LAST = create("last", null);
+
+        private String name;
+
+        private Phase(String name)
+        {
+            this.name = name;
+        }
+
+        /**
+         * @param name The name of your Phase. Used for profiling.
+         * @param chunkLayer If your Phase should render during LevelRenderer.renderChunkLayer, supply a RenderType. If not, pass null.
+         */
+        public static Phase create(ResourceLocation name, @Nullable RenderType chunkLayer)
```

Comment:
I can see how it would be interesting to create a `Phase` for a custom `RenderType` when something like https://github.com/MinecraftForge/MinecraftForge/pull/8331 is merged, but is there any reason for a mod to add a `Phase` that doesn't correspond to a `RenderType`? When and how would such a `Phase` be dispatched?

Reply:
A custom `Phase` would be an incredibly niche scenario in general, but ones not fired during a `RenderType` would likely be a "sub-Phase" of sorts, being called by a mod rendering during an existing `Phase` after doing some setup.